### PR TITLE
feat: Fast stdin display, lower memory usage, follow/reverse mode, home/end key support. long help support

### PR DESCRIPTION
**Pull request by chirino**
Maybe I got a little carried away.   It would have been better if this was done over several commits.  Feel free to cherry-pick ideas from this PR if you like any of its features.

The main reason I started down this path was that I expected `docker logs XXX -f | jlv` to behave similarly to `tail -f` with lo events just showing up as they occurred.  With this PR, stdin input is displayed instantly.  The log table now has a follow mode which is turned off once you navigate away from the latest log entry so the log entries will stop scrolling past the window.

While at it I noticed that we could optimize memory use by NOT holding log line data in memory.  We just need an offset, and length pair for each long line.  When the log is displayed we just load it back from the file.   Maybe we should increase the default MAX file size setting?  Might need some testing to figure out a better default.

If you have a fast log stream, it might be hard to get back into follow mode by pressing the up key enough times to go past the first item, so Home/End key support was added.

And while at it, I reversed log file order and option.  pressing r will toggle.

With so many new options, and mode, I added support for long help display.  Press ? to show the more advanced key binds.

### CODE REVIEW

**Review by hedhyw — COMMENTED** [review]
praise: thank you, I'll review it this week.

**Comment by hedhyw** [comment]
I'm still on it, need more time

File: internal/app/stateinitial_test.go [hunk 1]
```diff
@@ -19,12 +18,20 @@ import (
 func TestStateInitial(t *testing.T) {
 	t.Parallel()
 
+	is, err := source.Reader(bytes.NewReader([]byte{}), config.GetDefaultConfig())
```
Comment (hedhyw) [inline internal/app/stateinitial_test.go:21]: nitpick: `inputSource`.

File: internal/pkg/events/events.go [hunk 2]
```diff
@@ -37,9 +35,21 @@ func OpenJSONRowRequested(logEntries source.LazyLogEntries, index int) func() te
...
-// ViewRowsReloadRequested implements tea.Cmd. It creates ViewRowsReloadRequestedMsg.
-func ViewRowsReloadRequested() tea.Msg {
-	return ViewRowsReloadRequestedMsg{}
+func ShowError(err error) func() tea.Msg {
+	return func() tea.Msg {
+		return ErrorOccuredMsg{Err: err}
+	}
+}
+
+func HelpKeyClicked() tea.Msg {
```
Comment (hedhyw) [inline internal/pkg/events/events.go:44]: nitpick: can you please add comments to exported functions?

File: internal/pkg/source/source.go [hunk 3]
```diff
@@ -4,47 +4,143 @@ import (
...
 
-	bufReader := bufio.NewReaderSize(reader, maxLineSize)
-	logEntries := make(LazyLogEntries, 0, logEntriesEstimateNumber)
+	is.reader = bufio.NewReaderSize(io.LimitReader(is.file, is.maxSize), maxLineSize)
+	return is, nil
+}
+
+func Reader(input io.Reader, cfg *config.Config) (*Source, error) {
+	var err error
+	is := &Source{
```
Comment (hedhyw) [inline internal/pkg/source/source.go:60]: nitpick: `inputSource`

File: internal/pkg/source/sreamer.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+			entry, err := is.ReadLogEntry()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					if !is.CanFollow() {
+						return
+					}
+
+					// wait for new log entries to be written to the file,
+					// and try again.
+					ticker := time.NewTicker(200 * time.Millisecond)
```
Comment (hedhyw) [inline internal/pkg/source/sreamer.go:37]: question: why we don't use https://pkg.go.dev/time#NewTimer?
nitpick: let's create a `const` for `200 * time.Millisecond`.

File: internal/pkg/source/sreamer.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+)
+
+func (is *Source) StartStreaming(ctx context.Context, send func(msg LazyLogEntries, err error)) {
+	mu := sync.Mutex{}
```
Comment (hedhyw) [inline internal/pkg/source/sreamer.go:12]: nitpick: `logEntriesLock`, the function is big (`mu` is too generic).

File: internal/pkg/source/sreamer.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	logEntries := make([]LazyLogEntry, 0, 1000)
```
Comment (hedhyw) [inline internal/pkg/source/sreamer.go:13]: nitpick: let's create a `const` for `1000`.

File: internal/pkg/source/sreamer.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			entry, err := is.ReadLogEntry()
+			if err != nil {
```
Comment (hedhyw) [inline internal/pkg/source/sreamer.go:29]: nitpick: the code is too nested, let's think about way to simplify.

File: internal/app/lazytable.go [hunk 8]
```diff
@@ -10,56 +13,162 @@ import (
...
+		m = m.RenderedRows()
 	}
 
 	return m, cmd
 }
 
-func (m lazyTableModel[T]) withRenderedRows() lazyTableModel[T] {
-	cursor := m.table.Cursor()
+func (m lazyTableModel) handleKey(msg tea.KeyMsg, render bool) (lazyTableModel, bool) {
+	if key.Matches(msg, m.Application.keys.Reverse) {
```
Comment (hedhyw) [inline internal/app/lazytable.go:71]: suggestion: can you please write comments with explanation what's going on here, this function is quite complex to understand?

File: internal/pkg/source/source.go [hunk 9]
```diff
@@ -4,47 +4,143 @@ import (
...
-	logEntriesEstimateNumber = 256
 )
 
-// ParseLogEntriesFromReader reads the input and parses all logs.
-func ParseLogEntriesFromReader(
-	reader io.Reader,
-	cfg *config.Config,
-) (LazyLogEntries, error) {
-	reader = io.LimitReader(reader, cfg.MaxFileSizeBytes)
+type Source struct {
```
Comment (hedhyw) [inline internal/pkg/source/source.go:17]: suggestion: let's describe in comments how it works, the implementation is complex.

**Review by hedhyw — COMMENTED** [review]
- suggestion: there are many complex changes, can we extend tests? I can help with it, and yes, it's possible to do in another PR.
- note: I'll try to use it for a week for manual testing.
- todo: the PR title should be:
    "**feat:** Fast stdin display, lower memory usage, follow/reverse mode, home/end key support. long help support"
- todo: please also include "`Closes #52, #64, #79`" in the PR description.
- praise: nice, thank you for your pull request.

File: cmd/jlv/stdin_reader_mock.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+func getStdinReader(defaultInput fs.File) (io.Reader, error) {
+	r, w, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("pipe: %w", err)
+	}
+	go func() {
+		defer w.Close()
+		for i := 0; ; i++ {
+			_, err := w.Write([]byte(fmt.Sprintf(`{"message": "Line %d"}
+	`, i)))
```
[the remark this answers is not in the document]
Comment (chirino) [reply cmd/jlv/stdin_reader_mock.go:22]: no.. I was just using this mock stdin reader to be able to debug in my IDE in the stdin use case. (my ide did not allow me to pipe input while running under debug).

File: Makefile [hunk 11]
```diff
@@ -33,6 +33,11 @@ lint: bin/golangci-lint-${GOLANG_CI_LINT_VER}
 	./bin/golangci-lint-${GOLANG_CI_LINT_VER} run
 .PHONY: lint
 
+fix: bin/golangci-lint-${GOLANG_CI_LINT_VER}
+	 gofumpt -l -w .
```
[the remark this answers is not in the document]
Comment (chirino) [reply Makefile:37]: sounds good to me.

**Comment by chirino** [comment]
Rebased and applied your suggestions.  I'll be able to look into this more next week.

**Comment by chirino** [comment]
Fixed the linter warnings. Let me know if there are any other bits you want adjusted.

**Comment by hedhyw** [comment]
the test coverage decreased significantly, I can help with tests here

**Comment by hedhyw** [comment]
The latest changes don't work for me.

`make run`

image

commit **88a3750859b5421ad95585ed6fcd894f8134f01f**

File: internal/pkg/source/source.go [hunk 12]
```diff
@@ -4,47 +4,162 @@ import (
...
 
-		line = bytes.TrimSpace(line)
+func (is *Source) CanFollow() bool {
+	return len(is.name) != 0
+}
 
-		if len(line) > 0 {
-			lineClone := make([]byte, len(line))
-			copy(lineClone, line)
+var ErrFileTruncated = errors.New("file truncated")
```
Comment (hedhyw) [inline internal/pkg/source/source.go:119]:
```suggestion
const ErrFileTruncated semerr.Error = "file truncated"
```

**Comment by chirino** [comment]
That break is fixed now.

**Review by hedhyw — APPROVED** [review]

**Pull request merged**